### DISCUSSION TITLE
Too many connections: 1-1 correlation between request concurrency and connection pool size

### DISCUSSION THREAD

**Discussion by rcollette** [keycloak/keycloak#16122] [Q&A]
In load testing we are hitting a "Too many connections" error when our HTTP request concurrency reaches the configured number of database connections in the the pool via the KC_DB_POOL_MAX_SIZE setting.

Test Setup

- KC_DB_POOL_MAX size set to 30 in each of two pods in a kubernetes cluster
- AWS RDS MySql database with MAX_CONNECTIONS set to 64
- Three JMeter clients running 20 simultaneous threads obtaining new tokens or refreshing tokens.

We observe in the AWS console that the database active connections maxes out at 62 connections.
At peak we see that there are "Too many connections" errors logged by Keycloak, which on the one hand probably indicates we need to make the connection pool smaller even though it is already smaller than MAX_CONNECTIONS but I would expect see far more HTTP requests being handled than the number of database connections in the pool.

If we change the test to run three clients with 15 threads each we see no errors.

It appears as though perhaps database connections are being established for the lifetime of the HTTP request, rather than opening and closing the connection for just the operations that involve the database.   Typically database operations only require a small fraction of the overall request time and connections should be closed as soon as possible so that pooled connections can service many more HTTP requests than there are connections available.

Is there a way to improve this behavior to more effectively utilize available database connections?  I have searched all content for "Too many connections" and "pool" and have not been able to find any guidance other than a request to allow configuration of request/worker thread pooling in https://github.com/keycloak/keycloak/issues/15224fa

**Comment by anand-kulk** [comment]
@rcollette - did you find any workaround for the problem? we are getting same issue during load testing. keycloak is not respecting the --db-pool-max-size command line parameter.

**Reply by rcollette** [reply]
We are doing one more test, setting the pool size much lower than database max connections and will let you know.

**Reply by anand-kulk** [reply]
Unfortunately, that didn't work for us either. I created a separate thread as I am getting a bit different error from db, but reason is same - https://github.com/keycloak/keycloak/discussions/16153

**Reply by rcollette** [reply]
Actually, we just did a test setting the connection pool to 40 while the database supports over 60, and loaded it with three machines running 20 threads each. 

We no longer got sql connection errors and the connections to the server peaked at 40.  So it would appear that the connection pool configuration is being honored.

However, we were getting connection timeout errors on the client side and no errors on the server.   I would prefer to send some sort of error to the client when the server cannot respond in a reasonable amount of time, such as a 503 and get an error log for that.   With the current behavior, clients won't get serviced but we have no indication of a problem from the server.

**Reply by anand-kulk** [reply]
That is interesting, would you mind telling me what version of Quarkus keycloak you're using? Also which helm chart are you using to deploy?

**Reply by anand-kulk** [reply]
@rcollette, your trick of setting much lower number worked for me, I set max connections to 20 and I didn't get the pool error but I am getting timeout error `Sorry, acquisition timeout!`, is your timeout error also the same as mine?
I am running for 500 users/threads from a single JMeter client. My guess is there are very few connections in pool compared to threads (500) requesting connections and threads are timing out acquiring the connection from pool.

**Reply by rcollette** [reply]
It will be some time before I can respond.  US holiday till Tuesday and India offline at the moment.  My recollection is that we were seeing ConnectionTimeoutException in JMeter but JMeter is supposed to wait forever by default so I am trying to get more details about the test setup and the errors that occurred.
